sdks/python/apache_beam/runners/direct/sdf_direct_runner.py

```diff
@@ -0,0 +1,264 @@
+    self._process_element_invoker = process_element_invoker
+
+  def start_bundle(self):
+    # TODO: support start_bundle() method for SDFs.
+    pass
+
+  def finish_bundle(self):
+    # TODO: support finish_bundle() method for SDFs.
+    pass
+
+  def process(self, element, timestamp=beam.DoFn.TimestampParam,
+              window=beam.DoFn.WindowParam, * args, **kwargs):
+    if isinstance(element, KeyedWorkItem):
+      # Must be a timer firing.
+      key = element.encoded_key
+    else:
+      key, values = element
+      values = list(values)
+      assert len(values) == 1
+      value = values[0]
+
+    state = self._step_context.get_keyed_state(key)
+    element_state = state.get_state(window, self._element_tag)
+    is_seed_call = not element_state # Initially element_state is an empty list.
+
+    if not is_seed_call:
+      element = state.get_state(window, self._element_tag)
+      restriction = state.get_state(window, self._restriction_tag)
+      windowed_element = WindowedValue(element, timestamp, [window])
+    else:
+      element_and_restriction = (
+          value.value if isinstance(value, WindowedValue) else value)
+      assert isinstance(element_and_restriction, ElementAndRestriction)
+      element = element_and_restriction.element
+      restriction = element_and_restriction.restriction
+
+      if isinstance(value, WindowedValue):
+        windowed_element = WindowedValue(
+            element, value.timestamp, value.windows)
+      else:
+        windowed_element = WindowedValue(element, timestamp, [window])
+
+    tracker = self.sdf_invoker.invoke_create_tracker(restriction)
+    assert self._process_element_invoker
+    assert isinstance(self._process_element_invoker,
+                      SDFProcessElementInvoker)
+
+    output_values = self._process_element_invoker.invoke_process_element(
+        self.sdf_invoker, windowed_element, tracker)
+
+    sdf_result = None
+    for output in output_values:
+      if isinstance(output, SDFProcessElementInvoker.Result):
+        # SDFProcessElementInvoker.Result should be the last item yielded.
+        sdf_result = output
+        break
+      yield output
+
+    if not sdf_result.residual_restriction:
+      # All work for current residual and restriction pair is complete.
+      state.clear_state(window, self._element_tag)
+      state.clear_state(window, self._restriction_tag)
+      # Releasing output watermark by setting it to positive infinity.
+      state.add_state(window, self.watermark_hold_tag,
+                      WatermarkManager.WATERMARK_POS_INF)
+    else:
+      state.add_state(window, self._element_tag, element)
+      state.add_state(window, self._restriction_tag,
+                      sdf_result.residual_restriction)
+      state.add_state(window, self.watermark_hold_tag,
+                      WatermarkManager.WATERMARK_NEG_INF)
+      # Holding output watermark by setting it to negative infinity.
```

Comment:
<!--new_thread; commit:7549b5c2ebe2ae47af9066eaf97364a27e828ab5; resolved:0-->
This comment should be moved two lines up, to before the watermark hold call.